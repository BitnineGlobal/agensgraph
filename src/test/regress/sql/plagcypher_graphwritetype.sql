--
-- PL/agCypher - Use Parameter
--

-- setup

drop graph if exists g4 cascade;
drop function if exists func_create();
drop function if exists func_delete();
drop function if exists func_set();
drop function if exists func_merge();
drop function if exists func_remove();
drop function if exists func_complex1();
drop function if exists func_complex2();

create graph g4;
set graph_path=g4;

create (:person{name : 'Anders'}), (:person{name : 'Dilshad'}), (:person{name : 'Cesar'}), (:person{name : 'Becky'}), (:person{name : 'Filipa'}), (:person{name : 'Emil'});
create elabel block;

match (a), (b)
where a.name = 'Anders' and b.name = 'Dilshad'
create (a)-[e:knows{name:'friend1'}]->(b)
return e;

match (a), (b)
where a.name = 'Anders' and b.name = 'Cesar'
create (a)-[e:knows{name:'friend2'}]->(b)
return e;

match (a), (b)
where a.name = 'Anders' and b.name = 'Becky'
create (a)-[e:knows{name:'friend3'}]->(b)
return e;

match (a), (b)
where a.name = 'Dilshad' and b.name = 'Filipa'
create (a)-[e:knows{name:'friend4'}]->(b)
return e;

match (a), (b)
where a.name = 'Cesar' and b.name = 'Emil'
create (a)-[e:knows{name:'friend5'}]->(b)
return e;

match (a), (b)
where a.name = 'Becky' and b.name = 'Emil'
create (a)-[e:knows{name:'friend6'}]->(b)
return e;

-- test graphwrite type ( create )
create or replace function func_create() returns void as $$
begin
create( :person{name : 'Bossman'} );
end;
$$ language plagcypher;

match (a) return a;

select func_create();

match (a) return a;

create or replace function func_delete() returns void as $$
begin
match (a)
where a.name = 'Becky'
delete a;
end;
$$ language plagcypher;

match (a) return a;

select func_delete();

match (a) return a;

create or replace function func_set() returns void as $$
begin
match (a)
where a.name = 'Becky'
set a.name = 'lucy';
end;
$$ language plagcypher;

match (a) return a;

select func_set();

match (a) return a;

create or replace function func_merge() returns void as $$
begin
match (a) , (b)
where a.name = 'Cesar' and b.name = 'Filipa'
merge (a)-[e:block {name:'block'}]->(b);
end
$$ language plagcypher;

match (a)-[b]->(c)
where b.name = 'block'
return a , b , c;

select func_merge();

match (a)-[b]->(c)
where b.name = 'block'
return a , b , c;

create or replace function func_remove() returns void as $$
begin
match (a)
remove a.name;
end;
$$ language plagcypher;

match (a) return a;

select func_remove();

match (a) return a;

create vlabel v;
create elabel e;

create or replace function func_complex1() returns void as $$
declare
var1 edge;
var2 edge;
begin
match (a:v) , (b:v) , (c:v)
create (a)-[z:e {name:'edge1', id:'1'}]->(b)
create (b)-[r:e {name:'edge2', id:'2'}]->(c)
return z , r into var1 , var2;
end;
$$ language plagcypher;

select func_complex1();

create or replace function func_complex2() returns void as $$
declare
var1 edge;
begin
match (a:v) , (b:v)
merge (a)-[r:e {name:'edge' , id:'0'}]-(b)
on create set r.created = true, r.matched = null
on match set r.matched = true, r.created = null
return r into var1;
end;
$$ language plagcypher;

select func_complex2();

drop function if exists func_complex2();
drop function if exists func_complex1();
drop function if exists func_remove();
drop function if exists func_merge();
drop function if exists func_set();
drop function if exists func_delete();
drop function if exists func_create();
drop graph if exists g4 cascade;
