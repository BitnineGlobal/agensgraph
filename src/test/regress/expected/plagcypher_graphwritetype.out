--
-- PL/agCypher - Use Parameter
--
-- setup
DROP GRAPH IF EXISTS g4 CASCADE;
NOTICE:  graph "g4" does not exist, skipping
DROP FUNCTION IF EXISTS func_create();
NOTICE:  function func_create() does not exist, skipping
DROP FUNCTION IF EXISTS func_delete();
NOTICE:  function func_delete() does not exist, skipping
DROP FUNCTION IF EXISTS func_set();
NOTICE:  function func_set() does not exist, skipping
DROP FUNCTION IF EXISTS func_merge();
NOTICE:  function func_merge() does not exist, skipping
DROP FUNCTION IF EXISTS func_remove();
NOTICE:  function func_remove() does not exist, skipping
DROP FUNCTION IF EXISTS func_complex1();
NOTICE:  function func_complex1() does not exist, skipping
DROP FUNCTION IF EXISTS func_complex2();
NOTICE:  function func_complex2() does not exist, skipping
CREATE GRAPH g4;
SET GRAPH_PATH=g4;
CREATE ELABEL block;
CREATE (a:person{name : 'Anders'})-[:knows {name:'friend1'}]->(b:person{name : 'Dilshad'}),
(a)-[:knows {name:'friend2'}]->(c:person{name : 'Cesar'}),
(a)-[:knows {name:'friend3'}]->(d:person{name : 'Becky'}),
(b)-[:knows {name:'friend4'}]->(:person{name : 'Filipa'}),
(c)-[:knows {name:'friend5'}]->(e:person{name : 'Emil'}),
(d)-[:knows {name:'friend6'}]->(e);
-- test graphwrite type ( create )
CREATE OR REPLACE FUNCTION func_create() RETURNS void AS $$
BEGIN
CREATE( :person{name : 'Bossman'} );
END;
$$ LANGUAGE plagcypher;
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

SELECT func_create();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_create() line 3 at SQL statement
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

CREATE OR REPLACE FUNCTION func_delete() RETURNS void AS $$
BEGIN
MATCH (a)
WHERE a.name = 'Becky'
DELETE a;
END;
$$ LANGUAGE plagcypher;
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

SELECT func_delete();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_delete() line 3 at SQL statement
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

CREATE OR REPLACE FUNCTION func_set() RETURNS void AS $$
BEGIN
MATCH (a)
WHERE a.name = 'Becky'
SET a.name = 'lucy';
END;
$$ LANGUAGE plagcypher;
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

SELECT func_set();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_set() line 3 at SQL statement
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

CREATE OR REPLACE FUNCTION func_merge() RETURNS void AS $$
BEGIN
MATCH (a) , (b)
WHERE a.name = 'Cesar' AND b.name = 'Filipa'
MERGE (a)-[e:block {name:'block'}]->(b);
END;
$$ LANGUAGE plagcypher;
MATCH (a)-[b]->(c)
WHERE b.name = 'block'
RETURN a , b , c;
 a | b | c 
---+---+---
(0 rows)

SELECT func_merge();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_merge() line 3 at SQL statement
MATCH (a)-[b]->(c)
WHERE b.name = 'block'
RETURN a , b , c;
 a | b | c 
---+---+---
(0 rows)

CREATE OR REPLACE FUNCTION func_remove() RETURNS void AS $$
BEGIN
MATCH (a)
REMOVE a.name;
END;
$$ LANGUAGE plagcypher;
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

SELECT func_remove();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_remove() line 3 at SQL statement
MATCH (a) RETURN properties(a);
     properties      
---------------------
 {"name": "Anders"}
 {"name": "Dilshad"}
 {"name": "Cesar"}
 {"name": "Becky"}
 {"name": "Filipa"}
 {"name": "Emil"}
(6 rows)

CREATE VLABEL v;
CREATE ELABEL e;
CREATE OR REPLACE FUNCTION func_complex1() RETURNS void AS $$
DECLARE
var1 edge;
var2 edge;
BEGIN
MATCH (a:v) , (b:v) , (c:v)
CREATE (a)-[z:e {name:'edge1', id:'1'}]->(b)
CREATE (b)-[r:e {name:'edge2', id:'2'}]->(c)
RETURN z , r INTO var1 , var2;
END;
$$ LANGUAGE plagcypher;
SELECT func_complex1();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_complex1() line 6 at SQL statement
CREATE OR REPLACE FUNCTION func_complex2() RETURNS void AS $$
DECLARE
var1 edge;
BEGIN
MATCH (a:v) , (b:v)
MERGE (a)-[r:e {name:'edge' , id:'0'}]-(b)
ON CREATE SET r.created = true, r.matched = null
ON MATCH SET r.matched = true, r.created = null
RETURN r INTO var1;
END;
$$ LANGUAGE plagcypher;
SELECT func_complex2();
ERROR:  cannot modify graph in PL/pgSQL
CONTEXT:  PL/pgSQL function func_complex2() line 5 at SQL statement
DROP FUNCTION IF EXISTS func_complex2();
DROP FUNCTION IF EXISTS func_complex1();
DROP FUNCTION IF EXISTS func_remove();
DROP FUNCTION IF EXISTS func_merge();
DROP FUNCTION IF EXISTS func_set();
DROP FUNCTION IF EXISTS func_delete();
DROP FUNCTION IF EXISTS func_create();
DROP GRAPH IF EXISTS g4 CASCADE;
NOTICE:  drop cascades to 8 other objects
DETAIL:  drop cascades to sequence g4.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to elabel block
drop cascades to vlabel person
drop cascades to elabel knows
drop cascades to vlabel v
drop cascades to elabel e
